docs/docs/reference/dropped-features/package-objects.md

```diff
@@ -0,0 +1,44 @@
+---
+layout: doc-page
+title: "Dropped: Package Objects"
+---
+
+Package objects
+```scala
+    package object p {
+      val a = ...
+      def b = ...
+    }
+```
+will be dropped. They are still available in Scala 3.0, but will be deprecated and removed afterwards.
+
+Package objects are no longer needed since all kinds of definitions can now be written at the top-level. E.g.
+```scala
+    package p
+    type Labelled[T] = (String, T)
+    val a: Labelled[Int] = ("count", 1)
+    def b = a._2
+
+    case class C()
+
+    implicit object Cops {
+      def (x: C) pair (y: C) = (x, y)
+    }
+```
+There may be several source files in a package containing such toplevel definitions, and source files can freely mix toplevel value, method, and type definitions with classes and objects.
+
+The compiler generates synthetic objects that wrap toplevel definitions falling into one of the following categories:
+
+ - all pattern, value, method, and type definitions,
+ - implicit classes and objects,
+ - companion objects of opaque types.
```

Comment:
What about side-effecting expressions like `println("Hello")`? Those should work too right, since wrapping them in `val _ = println("Hello")` is trivial

Reply:
Yes, but the changes in parser do not support them. I am not sure we want to do this.

Reply:
If the goal is to try and normalize the various Scala source file syntaxes as @smarter wants, we would need to do this since all the various Scala scripting syntaxes (Ammonite's `amm`, `scala` command, Mill, SBT, worksheets, ...) all support top-level expressions like this.

This would also potentially make getting started with Scala much simpler: you could write code in a file, and run the file (this would require straightforward changes in the launcher script). Hello world could literally be `println("Hello World")` in a `Hello.scala`. That would move Scala closer to the Python getting started experience, I think in a good way.

Reply:
I agree with @lihaoyi-databricks about the benefits. I remember my first scala experience where I could do stuff in REPL that are impossible when I write a scala file and compile it (and I don't mean redeclaring defs and vals, of course).

Reply:
Maybe we should somehow separate the concept between an application and a library. Indeed a library shouldn't have side-effects values, but an application, why not? Why to write an application we need to extend `App` instead of just writing the contents in the file?

Reply:
@soronpo one neat way to disambiguate "applications" or "entrypoints" with top-level code and "libraries" without top-level code would be the file extension: `.sc` files are entrypoints, `.scala` files are library files. `.sc` is already supported by a lot of tools, editors such as IntelliJ, etc. which understand those files to allow top-level expressions.